Assets/Scripts/Framework/Train/TrainEngine.cs

```diff
@@ -0,0 +1,236 @@
+using FishNet.Object.Synchronizing;
+using System.Collections.Generic;
+using FishNet.Object;
+using DG.Tweening;
+using UnityEngine;
+using System;
+
+using DerailedDeliveries.Framework.Utils;
+
+namespace DerailedDeliveries.Framework.Train
+{
+    /// <summary>
+    /// Class responsible for controlling the trains engine.
+    /// </summary>
+    [RequireComponent(typeof(TrainController))]
+    public class TrainEngine : NetworkAbstractSingleton<TrainEngine>
+    {
+        [SerializeField]
+        private float _maxHighSpeed = 7.5f;
+
+        [SerializeField]
+        private float _maxMediumSpeed = 5f;
+
+        [SerializeField]
+        private float _maxLowSpeed = 2.5f;
+
+        [SerializeField, Space]
+        private float _accelerationDuration = 10f;
+
+        [SerializeField]
+        private Ease _accelerationEase = Ease.Linear;
+
+        /// <summary>
+        /// Current train engine state.
+        /// </summary>
+        public TrainEngineState EngineState 
+            { get; private set; } = TrainEngineState.OnStandby;
+
+        /// <summary>
+        /// Current train speed type.
+        /// </summary>
+        public TrainEngineSpeedTypes CurrentEngineSpeedType 
+            { get; private set; } = TrainEngineSpeedTypes.Still;
+
+        /// <summary>
+        /// Current train speed type.
+        /// </summary>
+        public TrainEngineSpeedTypes CurrentTargetEngineSpeedType 
+            { get; private set; } = TrainEngineSpeedTypes.Still;
+
+        /// <summary>
+        /// Current train velocity speed.
+        /// </summary>
+        public float CurrentVelocity => _currentSpeed / _trainController.SplineLength;
+
+        /// <summary>
+        /// Determines the chosen track for the next possible rail split.<br></br>
+        /// <br>False = left.</br>
+        /// <br>True = right.</br>
+        /// </summary>
+        public bool CurrentSplitDirection { get; set; }
+
+        /// <summary>
+        /// Invokes when train direction is changed.
+        /// </summary>
+        public Action<bool> OnDirectionChanged;
+
+        /// <summary>
+        /// Invoked when the train speed is changed.
+        /// </summary>
+        public Action<TrainEngineSpeedTypes> OnSpeedChanged;
+
+        /// <summary>
+        /// Invoked when the train target speed is changed.
+        /// </summary>
+        public Action<TrainEngineSpeedTypes> OnTargetSpeedChanged;
+
+        /// <summary>
+        /// Invoked when the train engine state is changed.
+        /// </summary>
+        public Action<TrainEngineState> OnEngineStateChanged;
+
+        private Dictionary<TrainEngineSpeedTypes, float> _speedValues;
+
+        [HideInInspector]
+        [SyncVar(Channel = FishNet.Transporting.Channel.Reliable)]
+        private float _currentSpeed;
+        
+        private float _speedTypesCount;
+
+        [SerializeField] private bool isLerping;
```

Comment:
Perhaps it's wise to move this up to the public and serialized vars due to it making more sense, unless all the vars around this are related in some way. Then it's up for discussion.

Reply:
This a test value which I currently use to debug the spam problem, this will be either deleted or kept private when I fix it an make it it spam-proof.

Reply:
Update: turned out that I didn't need this var, I deleted it.